STYLEGUIDE.md

```diff
@@ -0,0 +1,716 @@
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use types in the documentation */
+/**
+ * @returns {string} The description of the return value.
+ */
+
+/* Don't use @return */
+/**
+ * @return The description of the return value.
+ */
+
+/* Don't leave out the fullstop */
+/**
+ * @returns The description of the return value
+ */
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Apply all rules from above */
+/**
+ * @returns The description of the return value.
+ */
+```
+
+
+## Tests
+
+### Use .spec in test file names
+Use `.spec` not `.test` to differentiate test files.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use .test.ts */
+test-file.test.ts
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use .spec.ts */
+test-file.spec.ts
+```
+
+### Use it and describe notation
+Use `describe` and `it`, not `test`.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use test */
+test('Test Suite', () => {
+  test('Test Case', () => {
+    ...
+  });
+});
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use describe and it */
+describe('Test Suite', () => {
+  it('Test Case', () => {
+    ...
+  });
+});
+```
+
+### Put unit test files next to source files
+Create test files next to productive code, instead of a separate test directory.
+This makes it easier to keep the file structures in sync and find tests for a unit of code.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't create a separate test directory */
+src/
+  some-unit.ts
+test/
+  some-unit.spec.ts
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Create test files next to productive files */
+src/
+  some-unit.ts
+  some-unit.spec.ts
+```
+
+## Promises
+
+### Use `async`/`await` rather than `.then`
+In most cases the more modern `async`/`await` notation, makes code a bit easier to read, therefore this is the preferred way.
+However, there are some cases where `.then` is equally as good and potentially even less error-prone - in those cases both options are fine.
+This is especially the case if you have multiple asynchronous calls, that have common error handling (check the correct code examples).
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use .then by default */
+function getBar() {
+  return asyncWait().then(foo => foo.bar);
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use async/await as a default */
+async function getBar() {
+  return (await asyncWait()).bar;
+}
+
+/* Grey area: Feel free to use both variants when you have common error handling for multiple asynchronous calls */
+async function foo() {
+  try {
+    await asyncWait1();
+    return await asyncWait2(); // If you want to handle the error here, this await is necessary!
+  } catch(err) {
+    handleError(err);
+  }
+}
+
+/* Grey area: Feel free to use both variants when you have common error handling for multiple asynchronous calls */
+async function foo() {
+  return asyncWait1()
+    .then(() => asyncWait2())
+    .catch(err => handleError(err));
+}
```

Comment:
Should this also mention the async/await + `.catch` version?

```ts
async function foo() {
  return await asyncWait1().catch(err => handleError(err));
}
```

Reply:
Done, added an extra section.